consensus/polybft/validators_snapshot.go

```diff
@@ -36,127 +54,106 @@ func (v *validatorsSnapshotCache) GetSnapshot(blockNumber uint64, parents []*typ
 	v.lock.Lock()
 	defer v.lock.Unlock()
 
-	v.logger.Trace("Retrieving snapshot started...", "Block", blockNumber)
-
-	var snapshot AccountSet
-
-	lastBlockWithSnapshot := blockNumber
-
-	for ; ; lastBlockWithSnapshot-- {
-		snapshot = v.snapshots[lastBlockWithSnapshot]
-		if snapshot != nil {
-			v.logger.Trace("Found snapshot in memory cache", "Block", lastBlockWithSnapshot)
+	_, extra, err := getBlockData(blockNumber, v.blockchain)
+	if err != nil {
+		return nil, err
+	}
 
-			break
-		}
+	isEpochEndingBlock, err := v.isEpochEndingBlock(blockNumber, extra)
+	if err != nil {
+		return nil, err
+	}
 
-		dbSnapshot, err := v.state.getValidatorSnapshot(lastBlockWithSnapshot)
-		if err != nil {
-			return nil, fmt.Errorf(
-				"failed to retrieve validators snapshot from the database for block=%d: %w",
-				lastBlockWithSnapshot,
-				err,
-			)
-		}
+	epochToGetSnapshot := extra.Checkpoint.EpochNumber
+	if !isEpochEndingBlock {
+		epochToGetSnapshot--
+	}
 
-		if dbSnapshot != nil {
-			snapshot = dbSnapshot
-			v.snapshots[lastBlockWithSnapshot] = dbSnapshot.Copy()
-			v.logger.Trace("Found snapshot in database", "Block", lastBlockWithSnapshot)
+	v.logger.Trace("Retrieving snapshot started...", "Block", blockNumber, "Epoch", epochToGetSnapshot)
 
-			break
-		}
-
-		if lastBlockWithSnapshot == 0 {
-			// prevent underflow of uint64
-			break
-		}
+	latestValidatorSnapshot, err := v.getLastCachedSnapshot(epochToGetSnapshot)
+	if err != nil {
+		return nil, err
 	}
 
-	if snapshot != nil && lastBlockWithSnapshot == blockNumber {
-		if err := v.cleanup(); err != nil {
-			// error on cleanup should not block or fail any action
-			v.logger.Error("could not clean validator snapshots from cache and db", "err", err)
-		}
-
-		return snapshot, nil
+	if latestValidatorSnapshot != nil && latestValidatorSnapshot.Epoch == epochToGetSnapshot {
+		// we have snapshot for required block (epoch) in cache
+		return latestValidatorSnapshot.Snapshot, nil
 	}
 
-	if snapshot == nil {
-		// Haven't managed to retrieve snapshot for any block from the cache.
+	if latestValidatorSnapshot == nil {
+		// Haven't managed to retrieve snapshot for any epoch from the cache.
 		// Build snapshot from the scratch, by applying delta from the genesis block.
-		initialSnapshot, err := v.computeSnapshot(nil, 0, parents)
+		// log.Trace("Building validators snapshot from scratch...")
```

Comment:
Remove this log or change it a little bit (like "Building validators snapshot from genesis...") and uncomment it?